ext/standard/array.c

```diff
@@ -1976,6 +1976,14 @@ static void php_compact_var(HashTable *eg_active_symbol_table, zval *return_valu
 			ZVAL_COPY(&data, value_ptr);
 			zend_hash_update(Z_ARRVAL_P(return_value), Z_STR_P(entry), &data);
 		}
+		if (ZSTR_LEN(Z_STR_P(entry)) == sizeof("this")-1  && !strcmp(ZSTR_VAL(Z_STR_P(entry)), "this")) {
```

Comment:
`zend_string_equals_literal(Z_STR_P(entry), "this")` would be a bit more elegant here.